libraries/AP_WayBack/AP_WayBack.cpp

```diff
@@ -0,0 +1,818 @@
+            last_raw_point = p1;
+
+            last_loop_check = p0; // check new segments for intersections
+
+            if(left >1) {  // осталась одна новая точка 
+                last_reduce = move_back(last_big_reduce, SIMPLIFY_MOVE_BACK); // move back by SIMPLIFY_MOVE_BACK points
+                last_big_reduce = p1;
+                DBG_PRINTVAR( left); DBG_PRINTVARLN( last_reduce);
+
+                in_loop_reduce=true;   // check newly formed segments
+            } else { // no new points
+                last_reduce = p1; // just from last point
+            }
+
+
+    DBG_PRINTVARLN( left);
+
+    DBG_PRINTVAR( removed);   DBG_PRINTVAR( last_reduce);  DBG_PRINTVARLN( last_big_reduce);
+        
+            if((num_points-removed)>=max_num_points) goto again; // we still don't have a room for new point
+        }
+    }
+    
+    // if enabled or we should to clear room for new points: we check one leg per point so number of free points should be more than not checked legs
+    //                  process active    enabled                number of not checked legs          number of free points       corrected by one packet
+    if(! was_reduce ) {
+    
+        // to exclude quadratic complexity we check intersection only one step in time
+
+        if(in_loop_reduce){      //  уже обрабатываем или отстали на нужное количество точек 
+
+            // check one leg from last_loop_check
+            p0 = last_loop_check;
+            p1 = move_forw(p0,1); // points can be not removed so we use relative coordinates
+
+         // check segment for track intersection
+            uint16_t sb=0;
+            
+            if(p1>p0){
+    DBG_PRINT("check loop ");     DBG_PRINTVAR(p0);    DBG_PRINTVARLN(p1);
+
+                sb = try_remove_loop(p0, p1);
+            }
+
+
+            if(sb) { // loop found
+                p1=sb-1;    // prev point as last checked
+
+                last_reduce=p1; // try to reduce from here
+                new_points +=  MIN_SIMPLIFY_POINTS; // force reduce on next point
+            
+                if(p1<2) p1=2; // but not earlier than 2
+            
+            
+                was_reduce = true; // points was removed
+            } else {
+                if(p1 >= hi_loop_border)  in_loop_reduce=false;
+            }
+ 
+            
+            last_loop_check = p1;                    
+        }
+    }
+    // do it only if was any reductions
+    if(was_reduce) {
+        squizze();      // remove bad points
+    }
+
+}
+
+
+
+
+/*
+   Determine the intersection point of two line segments (Paul Bourke)
+   Return FALSE if the lines don't intersect
+*/
+bool AP_WayBack::linesIntersect(const Point &p1, const Point &p2,
+                                const Point &p3, const Point &p4,
+                                Point *cross) // coordinates of intersection
+{
+   float denom,numera,numerb;
+
+   denom  = (p4.y-p3.y) * (p2.x-p1.x) - (p4.x-p3.x) * (p2.y-p1.y);
```

Comment:
and in CortexM4F too :) GCC don't likes to optimize out classes and arrays, so stores data to memory even it can be calculated in registers. I *always* check assembly listings of my code.